### PR TITLE
node ip whitelist

### PR DESCRIPTION
**Pull request by interro**

### CODE REVIEW

File: settings_sample.py [hunk 1]
```diff
@@ -3,3 +3,6 @@
 DB_NAME = ''
 USER = ''
 PASSWD = ''
+NODE_WHITELIST_ENABLED = False
+# IP addresses separated by comma
+NODE_WHITELISTED_IP_ADDRESSES = ''
```
Comment (donce) [inline settings_sample.py]: Linter is complaining about missing newline. You can enable PyCharm to finish all lines automatically in preferences (ensure line feed...):

screen shot 2018-02-15 at 19 43 05

File: app.py [hunk 2]
```diff
@@ -93,6 +93,18 @@ def wrapper(*args, **kw):
     return wrapper
 
 
+def whitelisted_ip(f):
+    @wraps(f)
+    def wrapper(*args, **kw):
+        if settings.NODE_WHITELIST_ENABLED:
+            if request.remote_addr not in settings.NODE_WHITELISTED_IP_ADDRESSES.split(','):
```
Comment (donce) [inline app.py]: > ./app.py:100:80: E501 line too long (92 > 79 characters)

File: tests/test_app.py [hunk 3]
```diff
@@ -6,10 +6,13 @@
     generate_test_authorization,
     generate_static_public_address,
 )
+import settings
 
 
 class TestApi(TestCase):
     REMOTE_ADDR = '8.8.8.8'
+    settings.NODE_WHITELIST_ENABLED = False
+
```
Comment (donce) [inline tests/test_app.py]: > ./tests/test_app.py:17:5: E303 too many blank lines (2)

File: tests/test_app.py [hunk 4]
```diff
@@ -6,10 +6,13 @@
     generate_test_authorization,
     generate_static_public_address,
 )
+import settings
 
 
 class TestApi(TestCase):
     REMOTE_ADDR = '8.8.8.8'
+    settings.NODE_WHITELIST_ENABLED = False
```
Comment (donce) [inline tests/test_app.py]: It should be easy to test that decorator actually works - you can pass `remote_addr` param to `self._post` to test requests with different ips.

File: settings_sample.py (hunk 1 above)
Comment (donce) [inline settings_sample.py]: Why not a list?

```py
NODE_WHITELISTED_IP_ADDRESSES = [
  '123.123.123.123',
  '123.123.123.123',
  '123.123.123.123',
]
```

Comment (interro) [reply settings_sample.py]: thanks

File: app.py (hunk 2 above)
Comment (interro) [reply app.py]: done

File: tests/test_app.py (hunk 3 above)
Comment (interro) [reply tests/test_app.py]: done

File: settings_sample.py [hunk 5]
```diff
@@ -3,3 +3,6 @@
 DB_NAME = ''
 USER = ''
 PASSWD = ''
+NODE_WHITELIST_ENABLED = False
```
[the remark this answers is not in the document]
Comment (Waldz) [reply settings_sample.py]: see https://github.com/MysteriumNetwork/api/pull/29/files#diff-bdf3ecebd8379ca98cc89e545fc90899

File: tests/test_app.py (hunk 4 above)
Comment (interro) [reply tests/test_app.py]: tests now are there

File: app.py [hunk 6]
```diff
@@ -93,6 +93,19 @@ def wrapper(*args, **kw):
     return wrapper
 
 
+def whitelisted_ip(f):
```
[the remark this answers is not in the document]
Comment (interro) [reply app.py]: renamed

File: settings.py [hunk 7]
```diff
@@ -1,8 +1,18 @@
...
 DB_NAME = os.environ.get('DB_NAME')
 USER = os.environ.get('DB_USER')
 PASSWD = os.environ.get('DB_PASSWORD')
+
+# util.strtobool
+# True values are y, yes, t, true, on and 1;
+# False values are n, no, f, false, off and 0.
+# Raises ValueError if val is anything else.
+RESTRICT_BY_IP_ENABLED = util.strtobool(
+    os.environ.get('RESTRICT_BY_IP_ENABLED') or 'yes'
```
Comment (donce) [inline settings.py]: Disabling this feature by default would make more sense. It would work in development by default then.

File: settings.py [hunk 8]
```diff
@@ -1,8 +1,18 @@
...
+)
+ALLOWED_IP_ADDRESSES = os.environ.get('ALLOWED_IP_ADDRESSES') or ''
```
Comment (donce) [inline settings.py]: It's a bit inconsistent - `RESTRICT_BY_IP_ENABLED` is in python type (boolean), but `ALLOWED_IP_ADDRESSES` is not yet converted into python array.

They could be of the same level, i.e. both values raw values from ENV (two strings) or both values converted (boolean and array of strings)

File: settings.py (hunk 7 above)
Comment (interro) [reply settings.py]: Done

File: tests/test_app.py [hunk 9]
```diff
@@ -350,6 +358,7 @@ def test_ping_proposal(self):
         self.assertEqual({}, re.json)
 
     def test_node_send_stats_with_unknown_node(self):
+        settings.RESTRICT_BY_IP_ENABLED = False
```
Comment (donce) [inline tests/test_app.py]: All these disabling wouldn't be required, if this feature was disabled by default.

File: tests/test_app.py [hunk 10]
```diff
@@ -361,6 +370,42 @@ def test_node_send_stats_with_unknown_node(self):
         self.assertEqual(400, re.status_code)
         self.assertEqual({'error': 'node key not found'}, re.json)
 
+    def test_restrict_by_ip_fail(self):
+        settings.RESTRICT_BY_IP_ENABLED = True
+        settings.ALLOWED_IP_ADDRESSES = ''
```
Comment (donce) [inline tests/test_app.py]: I would add at least single ip address here to make sure that ips are compared - current tests can pass by implementing alternative simpler decorator implementation:
```py
if settings.ALLOWED_IP_ADDRESSES != '':
  # raise 403
```
You don't actually have to compare ips to pass these tests :)

File: settings.py (hunk 8 above)
Comment (interro) [reply settings.py]: converted ALLOWED_IP_ADDRESSES to array

File: tests/test_app.py (hunk 9 above)
Comment (interro) [reply tests/test_app.py]: removed

**Pull request merged**